src/com/reco1l/andengine/buffered/BufferedEntity.kt

```diff
@@ -0,0 +1,241 @@
+        }
+    }
+
+
+    //region Buffer lifecycle
+
+    protected open fun onRebuildBuffer(gl: GL10) = Unit
+
+    protected open fun onUpdateBuffer(gl: GL10, vararg data: Any) {
+        val buffer = buffer
+
+        if (buffer != null) {
+            buffer.update(gl, this, *data)
+
+            if (buffer is Buffer) {
+                buffer.setHardwareBufferNeedsUpdate()
+            }
+        }
+    }
+
+    //endregion
+
+    //region Draw pipeline
+
+    override fun onManagedDraw(gl: GL10, camera: Camera) {
+
+        if (shouldRebuildBuffer) {
+            shouldRebuildBuffer = false
+            onRebuildBuffer(gl)
+        }
+
+        if (shouldUpdateBuffer) {
+            shouldUpdateBuffer = false
+            onUpdateBuffer(gl)
+        }
+
+        super.onManagedDraw(gl, camera)
+    }
+
+    override fun doDraw(gl: GL10, camera: Camera) {
+        super.doDraw(gl, camera)
+        onDeclarePointers(gl)
+        onDrawBuffer(gl)
+    }
+
+    override fun beginDraw(gl: GL10) {
+        super.beginDraw(gl)
+
+        // Clearing
+        var clearMask = 0
+
+        if (clearInfo.depthBuffer) clearMask = clearMask or GL10.GL_DEPTH_BUFFER_BIT
+        if (clearInfo.colorBuffer) clearMask = clearMask or GL10.GL_COLOR_BUFFER_BIT
+        if (clearInfo.stencilBuffer) clearMask = clearMask or GL10.GL_STENCIL_BUFFER_BIT
+
+        if (clearMask != 0) {
+            gl.glClear(clearMask)
+        }
+
+        // Depth testing
+        if (depthInfo.test) {
+            gl.glDepthFunc(depthInfo.function)
+            gl.glDepthMask(depthInfo.mask)
+
+            GLHelper.enableDepthTest(gl)
+        } else {
+            GLHelper.disableDepthTest(gl)
+        }
+
+        // Blending
+        var sourceFactor = BLENDFUNCTION_SOURCE_DEFAULT
+        var destinationFactor = BLENDFUNCTION_DESTINATION_DEFAULT
+
+        if (blendInfo == BlendInfo.Inherit) {
+            val parent = parent
+
+            if (parent is BufferedEntity<*>) {
+                sourceFactor = parent.blendInfo.sourceFactor
+                destinationFactor = parent.blendInfo.destinationFactor
+            }
+        } else {
+            sourceFactor = blendInfo.sourceFactor
+            destinationFactor = blendInfo.destinationFactor
+        }
+
+        GLHelper.enableBlend(gl)
+        GLHelper.blendFunction(gl, sourceFactor, destinationFactor)
+
+        buffer?.beginDraw(gl)
+    }
+
+    protected open fun onDeclarePointers(gl: GL10) {
+        buffer?.declarePointers(gl, this)
+    }
+
+    protected open fun onDrawBuffer(gl: GL10) {
+        buffer?.draw(gl, this)
+    }
+
+    //endregion
+
+
+    override fun reset() {
+        super.reset()
+        blendInfo = BlendInfo.Mixture
+    }
+
+    fun finalize() {
+        buffer?.finalize()
+    }
+
+
+    companion object {
+
+        /**
+         * The buffer's data will be invalidated when the size of the entity changes.
+         */
+        const val InvalidateDataOnSizeChanged = 2
```

Comment:
Is there a reason why this starts at 2?

Reply:
There was a flag that started at 1 before